lib/http3/server.c

```diff
@@ -800,6 +800,7 @@ static void on_send_emit(quicly_stream_t *qs, size_t off, void *_dst, size_t *le
     /* retain the payload of response body before calling `h2o_proceed_request`, as the generator might discard the buffer */
     if (stream->state == H2O_HTTP3_SERVER_STREAM_STATE_SEND_BODY && *wrote_all &&
         quicly_sendstate_is_open(&stream->quic->sendstate) && !stream->proceed_requested) {
+        stream->req.timestamps.response_end_at = h2o_gettimeofday(stream->req.conn->ctx->loop);
```

Comment:
This is almost correct.

For consistency with http1.c, I think the exact condition should be when all of the following are met:
* `stream->state == H2O_HTTP3_SERVER_STREAM_STATE_SEND_BODY` (we are sending body)
* `*wrote_all` (we've written all we have in the buffer)
* `!quicly_sendstate_is_open(&stream->quic->sendstate)` (we will not receive more data from upstream, notice the not operator)
* `h2o_timeval_is_null(stream->req.timestamps.response_end_at)` (it is the first time to reach the above three conditions)

Reply:
I added @gfx 's test and made the changes above, however it does not seem like the send side gets closed? Is this because of the proxy setup? It does work correctly without the proxy setup.

Reply:
@sharksforarms Ah, I'm glad you noticed that. That is because when we forward a response by reading a socket, depending on the timing, the last element can be an EOS with an empty chunk. Then, the last QUIC STREAM frame generated would only carry EOS but any bytes. In such case, `on_send_emit` would not be called, because there is no payload to fulfill.

I think we have to move the logic of setting `response_end_at` to `scheduler_do_send` where we call `quicly_send_stream` for the bidirectional streams. There, we can determine if the stream is closed and if all the bytes have been sent at least once by comparing `size_inflight` and `final_size` of `st_h2o_http3_server_stream_t::quic->sendstate` (please refer to the doc-comments of these variables for rationale).

Could you please try that?

Reply:
Weird, tests pass in CI... are you able to re-run?

I get this locally. Based on my comment above, it does not seem to enter the else case of `quicly_sendstate_is_open`

```
# Subtest: HTTP/3
    ok 1 - http/3 is ok
    # { "protocol":"HTTP/3" , "connection-id":4 , "connect-time":0.004248 , "request-total-time":0 , "request-header-time":0 , "request-body-time":0 , "process-time":0.004739 , "response-time":null , "duration":null , "total-time":null , "error":"" , "proxy.idle-time":0 , "proxy.connect-time":0.000332 , "proxy.request-time":0 , "proxy.process-time":0.004407 , "proxy.response-time":0.103254 , "proxy.total-time":0.107661 }
    ok 2 - protocol
    ok 3
    ok 4
    ok 5
    ok 6
    ok 7
    ok 8
Use of uninitialized value $got in numeric ge (>=) at (eval in cmp_ok) t/50logconf.t line 162.
    not ok 9
    #   Failed test at t/50logconf.t line 162.
Use of uninitialized value $val in addition (+) at /usr/share/perl/5.32/Test/Builder.pm line 756.
    #     undef
    #         >=
    #     '0.1'
Use of uninitialized value $got in numeric ge (>=) at (eval in cmp_ok) t/50logconf.t line 163.
    not ok 10
    #   Failed test at t/50logconf.t line 163.
Use of uninitialized value $val in addition (+) at /usr/share/perl/5.32/Test/Builder.pm line 756.
    #     undef
    #         >=
    #     '0.1'
    ok 11
Use of uninitialized value $got in numeric ge (>=) at (eval in cmp_ok) t/50logconf.t line 165.
    not ok 12
    #   Failed test at t/50logconf.t line 165.
Use of uninitialized value $val in addition (+) at /usr/share/perl/5.32/Test/Builder.pm line 756.
    #     undef
    #         >=
    #     '0.1'
    ok 13
    ok 14
    ok 15
    ok 16
    ok 17
    ok 18
    1..18
    # Looks like you failed 3 tests of 18.
not ok 3 - HTTP/3
#   Failed test 'HTTP/3'
#   at t/50logconf.t line 172.
```

Reply:
> @sharksforarms Ah, I'm glad you noticed that. That is because when we forward a response by reading a socket, depending on the timing, the last element passed through `h2o_send` can be an EOS with an empty chunk (because at the socket level, EOS always arrives alone). Then, the last QUIC STREAM frame generated would only carry EOS but any bytes. In such case, `on_send_emit` would not be called, because there is no payload to be filled.
> 
> I think we have to move the logic of setting `response_end_at` to `scheduler_do_send` where we call `quicly_send_stream` for the bidirectional streams. There, we can determine if the stream is closed and if all the bytes have been sent at least once by comparing `size_inflight` and `final_size` of `st_h2o_http3_server_stream_t::quic->sendstate` (please refer to the doc-comments of these variables for rationale).
> 
> Could you please try that?

Thanks for explaining! 

This seems to work and tests pass locally now. Pending CI

Reply:
Now the CI fails with that error and local tests pass... hmm.

Reply:
For 5c160a7, I see CI green. Am I correct to believe that the problem has been fixed that was causing CI to fail?